src/conf_mode/protocols_failover.py

```diff
@@ -0,0 +1,116 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+
+from pathlib import Path
+
+from vyos.util import call
+from vyos.config import Config
+from vyos.configdict import dict_merge
+from vyos import ConfigError
+from vyos.template import render
+from vyos.xml import defaults
+from vyos import airbag
+
+airbag.enable()
+
+
+service_name = 'vyos-failover'
+service_conf = Path(f'/run/{service_name}.conf')
+systemd_service = '/etc/systemd/system/vyos-failover.service'
+rt_proto_failover = '/etc/iproute2/rt_protos.d/failover.conf'
+
+
+def get_config(config=None):
+    if config:
+        conf = config
+    else:
+        conf = Config()
+
+    base = ['protocols', 'failover']
+    failover = conf.get_config_dict(base, key_mangling=('-', '_'), get_first_key=True)
+
+    # Set default values only if we set config
+    if failover.get('route'):
+        for route, route_config in failover.get('route').items():
+            for next_hop, next_hop_config in route_config.get('next_hop').items():
+                default_values = defaults(base + ['route'])
+                failover['route'][route]['next_hop'][next_hop] = dict_merge(
+                    default_values['next_hop'], failover['route'][route]['next_hop'][next_hop])
+
+    return failover
+
+def verify(failover):
+    # bail out early - looks like removal from running config
+    if not failover:
+        return None
+
+    if 'route' not in failover:
+        raise ConfigError(f'Failover "route" is mandatory!')
+
+    for route, route_config in failover['route'].items():
+        if not route_config.get('next_hop'):
+            raise ConfigError(f'Next-hop for "{route}" is mandatory!')
+
+        for next_hop, next_hop_config in route_config.get('next_hop').items():
+            if 'interface' not in next_hop_config:
+                raise ConfigError(f'Interface for route "{route}" next-hop "{next_hop}" is mandatory!')
+
+            if not next_hop_config.get('check'):
+                raise ConfigError(f'Check target for next-hop "{next_hop}" is mandatory!')
+
+            if 'target' not in next_hop_config['check']:
+                raise ConfigError(f'Check target for next-hop "{next_hop}" is mandatory!')
+    return None
+
+def generate(failover):
+    if not failover:
+        service_conf.unlink(missing_ok=True)
+        return None
+
+    # Add own rt_proto 'failover'
+    # Helps to detect all own routes 'proto failover'
+    with open(rt_proto_failover, 'w') as f:
+        f.write('111  failover\n')
+
+    # Write configuration file
+    conf_json = json.dumps(failover, indent=4)
```

Comment:
Myself I'm inclined to use TOML for configs. It's not any harder to use, but much easier to read.

Reply:
I like more JSON format:
```
{
    "route": {
        "203.0.113.1/32": {
            "next_hop": {
                "192.168.122.1": {
                    "check": {
                        "target": "192.168.122.1",
                        "timeout": "10",
                        "type": "icmp"
                    },
                    "interface": "eth0",
                    "metric": "2"
                }
            }
        },
        "203.0.113.2/32": {
            "next_hop": {
                "192.168.122.1": {
                    "check": {
                        "target": "192.168.122.1",
                        "timeout": "10",
                        "type": "icmp"
                    },
                    "interface": "eth0",
                    "metric": "2"
                }
            }
        },
        "203.0.113.3/32": {
            "next_hop": {
                "192.168.122.1": {
                    "check": {
                        "target": "192.168.122.1",
                        "timeout": "10",
                        "type": "icmp"
                    },
                    "interface": "eth0",
                    "metric": "2"
                }
            }
        },
        "203.0.113.4/32": {
            "next_hop": {
                "192.168.122.1": {
                    "check": {
                        "target": "192.168.122.1",
                        "timeout": "10",
                        "type": "icmp"
                    },
                    "interface": "eth0",
                    "metric": "2"
                }
            }
        },
        "203.0.113.5/32": {
            "next_hop": {
                "192.168.122.1": {
                    "check": {
                        "target": "192.168.122.1",
                        "timeout": "10",
                        "type": "icmp"
                    },
                    "interface": "eth0",
                    "metric": "2"
                }
            }
        }
    }
}

```
TOML the same config:
```
[route."203.0.113.1/32".next_hop."192.168.122.1"]
interface = "eth0"
metric = "2"

[route."203.0.113.2/32".next_hop."192.168.122.1"]
interface = "eth0"
metric = "2"

[route."203.0.113.3/32".next_hop."192.168.122.1"]
interface = "eth0"
metric = "2"

[route."203.0.113.4/32".next_hop."192.168.122.1"]
interface = "eth0"
metric = "2"

[route."203.0.113.5/32".next_hop."192.168.122.1"]
interface = "eth0"
metric = "2"

[route."203.0.113.1/32".next_hop."192.168.122.1".check]
target = "192.168.122.1"
timeout = "10"
type = "icmp"

[route."203.0.113.2/32".next_hop."192.168.122.1".check]
target = "192.168.122.1"
timeout = "10"
type = "icmp"

[route."203.0.113.3/32".next_hop."192.168.122.1".check]
target = "192.168.122.1"
timeout = "10"
type = "icmp"

[route."203.0.113.4/32".next_hop."192.168.122.1".check]
target = "192.168.122.1"
timeout = "10"
type = "icmp"

[route."203.0.113.5/32".next_hop."192.168.122.1".check]
target = "192.168.122.1"
timeout = "10"
type = "icmp"

```
As for me, JSON is more readable in that case.